### ISSUE TITLE
Implement fast-nms

### ISSUE THREAD

**Issue by fsx950223** [tensorflow/addons#671]
**Describe the feature and the current behavior/state.**
The fast-nms algorithm has better performance than traditional nms.
Described in [paper](https://arxiv.org/pdf/1904.02689.pdf).
Kernel implementation may provides better performance than python implementation.
**Relevant information**
- Are you willing to contribute it (yes/no): yes
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where): https://arxiv.org/pdf/1904.02689.pdf
- Is there already an implementation in another framework? (if so, where): https://github.com/dbolya/yolact/blob/1722387d75210361c1f21c911d0e2420a48c7a23/layers/functions/detection.py#L136
- Was it part of tf.contrib? (if so, where): no

**Which API type would this fall under (layer, metric, optimizer, etc.)**

**Who will benefit with this feature?**

**Any other info.**
Tensorflow Implementation on [colab](https://colab.research.google.com/drive/1LhrD586OYojS_qXf309_5d2hx1QapFqk)

**Comment by SSaishruthi** [comment]
Is anyone working on this? If not, I will take a look @seanpmorgan

**Comment by seanpmorgan** [comment]
> Is anyone working on this? If not, I will take a look @seanpmorgan

Not that I'm aware of, but @fsx950223 said they are willing to implement. If there has been no progress then should be fine to start on.

**Comment by fsx950223** [comment]
I'm a little busy for now, I will try it next week. Thanks.

**Comment by SSaishruthi** [comment]
I will wait for the next update on this.

**Comment by samikama** [comment]
@fsx950223 how does this algorithm compare against existing GPU NMS implementations in TF? Do you have any idea?

**Comment by fsx950223** [comment]
> @fsx950223 how does this algorithm compare against existing GPU NMS implementations in TF? Do you have any idea?

Tensorflow benchmark could compare performance between them.

**Comment by bhack** [comment]
@seanpmorgan We have an upstream fast nms kernel in Tensorflow Lite https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/detection_postprocess.cc#L456-L463. @seanpmorgan I think this could be more in the "perimeter" of keras-cv (/cc @tanzhenyu ) as we have close yesterday the other image processing related issues/PRs.

**Comment by AutumnZ-94** [comment]
BTW, `tf.image.non_max_suppression` performs 20ms slower already than `tf.image.non_max_suppression_with_scores` as I tested.

**Comment by anshkumar** [comment]
@bhack  I'm not sure how fast nms kernel in Tensorflow Lite will work in normal scenario. Anyway, it's not hard to implement fast-nms. 
```
def _cc_fast_nms(self, boxes, masks, scores, iou_threshold:float=0.5, top_k:int=15):
        # Cross Class NMS
        # Collapse all the classes into 1 
        classes = tf.argmax(scores, axis=-1)+1
        scores = tf.reduce_max(scores, axis=-1)
        _, idx = tf.math.top_k(scores, k=tf.math.minimum(top_k, tf.shape(scores)[0]))
        boxes_idx = tf.gather(boxes, idx, axis=0)

        # Compute the pairwise IoU between the boxes
        iou = utils._iou(boxes_idx, boxes_idx)

        # Zero out the lower triangle of the cosine similarity matrix and diagonal
        iou = tf.linalg.band_part(iou, 0, -1) - tf.linalg.band_part(iou, 0, 0)

        # Now that everything in the diagonal and below is zeroed out, if we take the max
        # of the IoU matrix along the columns, each column will represent the maximum IoU
        # between this element and every element with a higher score than this element.
        iou_max = tf.reduce_max(iou, axis=0)

        # Now just filter out the ones greater than the threshold, i.e., only keep boxes that
        # don't have a higher scoring box that would supress it in normal NMS.
        idx_out = idx[iou_max <= iou_threshold]

        classes = tf.cast(tf.gather_nd(classes, tf.expand_dims(idx_out, axis=-1)), dtype=tf.float32)
        boxes = tf.gather_nd(boxes, tf.expand_dims(idx_out, axis=-1))
        masks = tf.gather_nd(masks, tf.expand_dims(idx_out, axis=-1))
        scores = tf.gather_nd(scores, tf.expand_dims(idx_out, axis=-1))

        return boxes, masks, classes, scores
```
